### PR TITLE
Separate eager loaded relations for fields with aliases

### PR DESCRIPTION
**Pull request by spawnia**
- [x] Added or updated tests
- [x] Documented user facing changes
- [x] Updated CHANGELOG.md

**Changes**

Fixes a bug in relation batch loading where querying the same relation in multiple aliased fields would conflate the results.

**Breaking changes**

None.

### CODE REVIEW

**Comment by spawnia** [comment]
Even though the relation is loaded twice with different children, those actions are resolved sequentially. This is due to Lighthouse creating two different batch loaders because of the aliased path.

image

File: src/Execution/ModelsLoader/SimpleModelsLoader.php [hunk 1]
```diff
@@ -18,15 +19,35 @@ class SimpleModelsLoader implements ModelsLoader
...
+     *
+     * @var string
+     */
+    protected $key;
+
     public function __construct(string $relation, Closure $decorateBuilder)
     {
         $this->relation = $relation;
         $this->decorateBuilder = $decorateBuilder;
+        $this->key = Str::uuid()->toString();
```
[the remark this answers is not in the document]
Comment (olivernybroe) [reply src/Execution/ModelsLoader/SimpleModelsLoader.php:33]: You are saying sufficiently unique, but isn't the path unique per query? 🤔

[the remark this answers is not in the document]
Comment (olivernybroe) [reply src/Execution/ModelsLoader/SimpleModelsLoader.php:33]: As this is an issue that prob. is in all the models loaders, maybe the key part should happen before the modes loader is doing it's work 🤔

File: src/Execution/ModelsLoader/SimpleModelsLoader.php [hunk 2]
```diff
@@ -18,15 +19,35 @@ class SimpleModelsLoader implements ModelsLoader
...
+        // We just return the first level of relations for now.
+        // They hold the nested relations in case they are needed.
+        $firstRelation = $parts[0];
+
         $parents->load([$this->relation => $this->decorateBuilder]);
+
+        foreach ($parents as $model) {
+            $relation = $model->getRelation($firstRelation);
+            $model->setRelation($this->key, $relation);
+        }
```
Comment (olivernybroe) [inline src/Execution/ModelsLoader/SimpleModelsLoader.php:50]: Not sure I get this loop 🤔 
It seems like we keep setting the same relation as we are doing it based on the key. If this is the desired result, can't we just take the last element in the parents array and set that?

Comment (spawnia) [reply src/Execution/ModelsLoader/SimpleModelsLoader.php:50]: Totally rewrote it.

File: src/Execution/ModelsLoader/SimpleModelsLoader.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (spawnia) [reply src/Execution/ModelsLoader/SimpleModelsLoader.php:33]: Good point, I figured out a solution that is external to the `ModelsLoader` and requires very little trickery.

**Comment by spawnia** [comment]
@olivernybroe Can you try out if this version fixes the specific issue in your application and otherwise works for you?

**Review by olivernybroe — APPROVED** [review]
Nice! When using this branch, the bug we experienced with it is fixed 👍

**Pull request merged**